File: src/c47/distributions/weibull.c
/* This file is part of 43S.
 *
 * 43S is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * 43S is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with 43S.  If not, see <http://www.gnu.org/licenses/>.
 */

/********************************************//**
 * \file weibull.c
 ***********************************************/

#include "distributions/weibull.h"

#include "constantPointers.h"
#include "error.h"
#include "flags.h"
#include "fonts.h"
#include "mathematics/comparisonReals.h"
#include "mathematics/wp34s.h"
#include "registers.h"
#include "registerValueConversions.h"

#include "c47.h"


#if defined(SAVE_SPACE_DM42_15)
  void fnWeibullP     (uint16_t unusedButMandatoryParameter){}
  void fnWeibullL     (uint16_t unusedButMandatoryParameter){}
  void fnWeibullR     (uint16_t unusedButMandatoryParameter){}
  void fnWeibullI     (uint16_t unusedButMandatoryParameter){}
  void WP34S_Pdf_Weib (const real_t *x, const real_t *b, const real_t *t, real_t *res, realContext_t *realContext){}
  void WP34S_Cdfu_Weib(const real_t *x, const real_t *b, const real_t *t, real_t *res, realContext_t *realContext){}
  void WP34S_Cdf_Weib (const real_t *x, const real_t *b, const real_t *t, real_t *res, realContext_t *realContext){}
  void WP34S_Qf_Weib  (const real_t *x, const real_t *b, const real_t *t, real_t *res, realContext_t *realContext){}

#else
  static bool_t checkParamWeibull(real_t *x, real_t *i, real_t *j) {
    if (!getRegisterAsReal(REGISTER_X, x)
        || !getRegisterAsReal(REGISTER_M, i)
        || !getRegisterAsReal(REGISTER_S, j))
        goto err;

    if(realIsNegative(x)) {
      displayDomainErrorMessage(ERROR_ARG_EXCEEDS_FUNCTION_DOMAIN, ERR_REGISTER_LINE, REGISTER_X);
      #if(EXTRA_INFO_ON_CALC_ERROR == 1)
        moreInfoOnError("In function checkParamWeibull:", "cannot calculate for x < 0", NULL, NULL);
      #endif // (EXTRA_INFO_ON_CALC_ERROR == 1)
      goto err;
    }
    else if(realIsZero(i) || realIsNegative(i) || realIsZero(j) || realIsNegative(j)) {
      displayDomainErrorMessage(ERROR_INVALID_DISTRIBUTION_PARAM, ERR_REGISTER_LINE, REGISTER_X);
      #if(EXTRA_INFO_ON_CALC_ERROR == 1)
        moreInfoOnError("In function checkParamWeibull:", "cannot calculate for b " STD_LESS_EQUAL " 0 or t " STD_LESS_EQUAL " 0", NULL, NULL);
      #endif // (EXTRA_INFO_ON_CALC_ERROR == 1)
      goto err;
    }
    return true;

  err:
    if(getSystemFlag(FLAG_SPCRES)) {
      convertRealToReal34ResultRegister(const_NaN, REGISTER_X);
    }
    return false;
  }


  void fnWeibullP(uint16_t unusedButMandatoryParameter) {
    real_t val, shape, lifetime, ans;

    if(!saveLastX()) {
      return;
    }

    if(checkParamWeibull(&val, &shape, &lifetime)) {
      WP34S_Pdf_Weib(&val, &shape, &lifetime, &ans, &ctxtReal39);
      reallocateRegister(REGISTER_X, dtReal34, REAL34_SIZE_IN_BLOCKS, amNone);
      convertRealToReal34ResultRegister(&ans, REGISTER_X);
    }

    adjustResult(REGISTER_X, false, false, REGISTER_X, -1, -1);
  }

  void fnWeibullL(uint16_t unusedButMandatoryParameter) {
    real_t val, shape, lifetime, ans;

    if(!saveLastX()) {
      return;
    }

    if(checkParamWeibull(&val, &shape, &lifetime)) {
      WP34S_Cdf_Weib(&val, &shape, &lifetime, &ans, &ctxtReal39);
      reallocateRegister(REGISTER_X, dtReal34, REAL34_SIZE_IN_BLOCKS, amNone);
      convertRealToReal34ResultRegister(&ans, REGISTER_X);
    }

    adjustResult(REGISTER_X, false, false, REGISTER_X, -1, -1);
  }

  void fnWeibullR(uint16_t unusedButMandatoryParameter) {
    real_t val, shape, lifetime, ans;

    if(!saveLastX()) {
      return;
    }

    if(checkParamWeibull(&val, &shape, &lifetime)) {
      WP34S_Cdfu_Weib(&val, &shape, &lifetime, &ans, &ctxtReal39);
      reallocateRegister(REGISTER_X, dtReal34, REAL34_SIZE_IN_BLOCKS, amNone);
      convertRealToReal34ResultRegister(&ans, REGISTER_X);
    }

    adjustResult(REGISTER_X, false, false, REGISTER_X, -1, -1);
  }

  void fnWeibullI(uint16_t unusedButMandatoryParameter) {
    real_t val, shape, lifetime, ans;

    if(!saveLastX()) {
      return;
    }

    if(checkParamWeibull(&val, &shape, &lifetime)) {
      if(realCompareLessEqual(&val, const_0) || realCompareGreaterEqual(&val, const_1)) {
        displayDomainErrorMessage(ERROR_ARG_EXCEEDS_FUNCTION_DOMAIN, ERR_REGISTER_LINE, REGISTER_X);
        #if(EXTRA_INFO_ON_CALC_ERROR == 1)
          moreInfoOnError("In function fnWeibullI:", "the argument must be 0 < x < 1", NULL, NULL);
        #endif // (EXTRA_INFO_ON_CALC_ERROR == 1)
        if(getSystemFlag(FLAG_SPCRES)) {
          convertRealToReal34ResultRegister(const_NaN, REGISTER_X);
        }
        return;
      }
      WP34S_Qf_Weib(&val, &shape, &lifetime, &ans, &ctxtReal39);
      reallocateRegister(REGISTER_X, dtReal34, REAL34_SIZE_IN_BLOCKS, amNone);
      convertRealToReal34ResultRegister(&ans, REGISTER_X);
    }

    adjustResult(REGISTER_X, false, false, REGISTER_X, -1, -1);
  }



  /******************************************************
   * This functions are borrowed from the WP34S project
   ******************************************************/

  void WP34S_Pdf_Weib(const real_t *x, const real_t *b, const real_t *t, real_t *res, realContext_t *realContext) {
    real_t p, q, r;

    realDivide(x, b, &p, realContext);
    if(realIsSpecial(&p) || realIsNegative(&p) || realIsZero(&p)) {
      realZero(res);
      return;
    }
    realPower(&p, t, &q, realContext);
    realMultiply(&q, const__1, &r, realContext);
    realExp(&r, &r, realContext);
    realMultiply(&r, &q, &r, realContext);
    realDivide(&r, &p, &r, realContext);
    realMultiply(&r, t, &r, realContext);
    realDivide(&r, b, res, realContext);
  }

  void WP34S_Cdfu_Weib(const real_t *x, const real_t *b, const real_t *t, real_t *res, realContext_t *realContext) {
    real_t p;

    realDivide(x, b, &p, realContext);
    if(realIsNegative(&p) || realIsZero(&p)) {
      realCopy(const_1, res);
      return;
    }
    if(realIsSpecial(&p)) {
      realZero(res);
      return;
    }
    realPower(&p, t, &p, realContext);
    realChangeSign(&p);
    realExp(&p, res, realContext);
  }

  void WP34S_Cdf_Weib(const real_t *x, const real_t *b, const real_t *t, real_t *res, realContext_t *realContext) {
    real_t p;

    realDivide(x, b, &p, realContext);
    if(realIsNegative(&p) || realIsZero(&p)) {
      realZero(res);
      return;
    }
    if(realIsSpecial(&p)) {
      realCopy(const_1, res);
      return;
    }
    realPower(&p, t, &p, realContext);
    realChangeSign(&p);
    WP34S_ExpM1(&p, res, realContext);
    realChangeSign(res);
  }

  void WP34S_Qf_Weib(const real_t *x, const real_t *b, const real_t *t, real_t *res, realContext_t *realContext) {
    /* (-ln(1-p) ^ (1/k)) * J */
    real_t p, q;

    realMultiply(x, const__1, &p, realContext);
    WP34S_Ln1P(&p, &p, realContext);
    realChangeSign(&p);
    realDivide(const_1, t, &q, realContext);
    realPower(&p, &q, &p, realContext);
    realMultiply(&p, b, res, realContext);
  }

#endif //SAVE_SPACE_DM42_15

File: zig_src/shell/distributions/weibull.zig
// SPDX-License-Identifier: GPL-3.0-only
//
// Zig owner for the Weibull distribution commands, porting
// `src/c47/distributions/weibull.c`. The four fn* entry points are dispatched
// from the items.c function table (exported via frontier.zig); the WP34S_*_Weib
// helpers are weibull-internal and stay private. All decNumber/register access
// goes through frontier_distribution_runtime.

const dr = @import("distribution_runtime.zig");
const real_t = dr.real_t;
const realContext_t = dr.realContext_t;

fn checkParamWeibull(x: *real_t, shape: *real_t, scale: *real_t) linksection(dr.code_section) bool {
    if (!dr.saveLastX()) {
        return false;
    }

    if (!dr.getRegisterAsReal(dr.REGISTER_X, x) or !dr.getRegisterAsReal(dr.REGISTER_Q, shape) or !dr.getRegisterAsReal(dr.REGISTER_S, scale)) {
        dr.specialResultNaN();
        return false;
    }

    if (dr.realIsNegative(x)) {
        dr.displayDomainErrorMessage(dr.ERROR_ARG_EXCEEDS_FUNCTION_DOMAIN, dr.ERR_REGISTER_LINE, dr.REGISTER_X);
        dr.moreInfoOnError("In function checkParamWeibull:", "cannot calculate for x < 0", null, null);
        dr.specialResultNaN();
        return false;
    } else if (dr.realIsZero(shape) or dr.realIsNegative(shape) or dr.realIsZero(scale) or dr.realIsNegative(scale)) {
        dr.displayDomainErrorMessage(dr.ERROR_INVALID_DISTRIBUTION_PARAM, dr.ERR_REGISTER_LINE, dr.REGISTER_X);
        dr.moreInfoOnError("In function checkParamWeibull:", "cannot calculate for b <= 0 or t <= 0", null, null);
        dr.specialResultNaN();
        return false;
    }
    return true;
}

pub fn weibullP(unused_but_mandatory_parameter: u16) linksection(dr.code_section) void {
    _ = unused_but_mandatory_parameter;
    var val: real_t = undefined;
    var shape: real_t = undefined;
    var lifetime: real_t = undefined;
    var ans: real_t = undefined;

    if (checkParamWeibull(&val, &shape, &lifetime)) {
        wp34sPdfWeib(&val, &lifetime, &shape, &ans, &dr.ctxtReal39);
        dr.storeResult(&ans);
    }
}

pub fn weibullL(unused_but_mandatory_parameter: u16) linksection(dr.code_section) void {
    _ = unused_but_mandatory_parameter;
    var val: real_t = undefined;
    var shape: real_t = undefined;
    var lifetime: real_t = undefined;
    var ans: real_t = undefined;

    if (checkParamWeibull(&val, &shape, &lifetime)) {
        wp34sCdfWeib(&val, &lifetime, &shape, &ans, &dr.ctxtReal39);
        dr.storeResult(&ans);
    }
}

pub fn weibullR(unused_but_mandatory_parameter: u16) linksection(dr.code_section) void {
    _ = unused_but_mandatory_parameter;
    var val: real_t = undefined;
    var shape: real_t = undefined;
    var lifetime: real_t = undefined;
    var ans: real_t = undefined;

    if (checkParamWeibull(&val, &shape, &lifetime)) {
        wp34sCdfuWeib(&val, &lifetime, &shape, &ans, &dr.ctxtReal39);
        dr.storeResult(&ans);
    }
}

pub fn weibullI(unused_but_mandatory_parameter: u16) linksection(dr.code_section) void {
    _ = unused_but_mandatory_parameter;
    var val: real_t = undefined;
    var shape: real_t = undefined;
    var lifetime: real_t = undefined;
    var ans: real_t = undefined;

    if (checkParamWeibull(&val, &shape, &lifetime)) {
        if (dr.realCompareLessEqual(&val, dr.const0()) or dr.realCompareGreaterEqual(&val, dr.const1())) {
            dr.displayDomainErrorMessage(dr.ERROR_ARG_EXCEEDS_FUNCTION_DOMAIN, dr.ERR_REGISTER_LINE, dr.REGISTER_X);
            dr.moreInfoOnError("In function fnWeibullI:", "the argument must be 0 < x < 1", null, null);
            dr.specialResultNaN();
            return;
        }
        wp34sQfWeib(&val, &lifetime, &shape, &ans, &dr.ctxtReal39);
        dr.storeResult(&ans);
    }
}

// These functions are borrowed from the WP34S project.

fn wp34sPdfWeib(x: *const real_t, b: *const real_t, t: *const real_t, res: *real_t, real_context: *realContext_t) linksection(dr.code_section) void {
    var p: real_t = undefined;
    var q: real_t = undefined;
    var r: real_t = undefined;

    dr.realDivide(x, b, &p, real_context);
    if (dr.realIsSpecial(&p) or dr.realIsNegative(&p) or dr.realIsZero(&p)) {
        dr.setZero(res);
        return;
    }
    dr.realPow(&p, t, &q, real_context);
    dr.realMinus(&q, &r, real_context);
    dr.realExponential(&r, &r, real_context);
    dr.realMultiply(&r, &q, &r, real_context);
    dr.realDivide(&r, &p, &r, real_context);
    dr.realMultiply(&r, t, &r, real_context);
    dr.realDivide(&r, b, res, real_context);
}

fn wp34sCdfuWeib(x: *const real_t, b: *const real_t, t: *const real_t, res: *real_t, real_context: *realContext_t) linksection(dr.code_section) void {
    var p: real_t = undefined;

    dr.realDivide(x, b, &p, real_context);
    if (dr.realIsNegative(&p) or dr.realIsZero(&p)) {
        dr.setOne(res);
        return;
    }
    if (dr.realIsSpecial(&p)) {
        dr.setZero(res);
        return;
    }
    dr.realPow(&p, t, &p, real_context);
    dr.realChangeSign(&p);
    dr.realExponential(&p, res, real_context);
}

fn wp34sCdfWeib(x: *const real_t, b: *const real_t, t: *const real_t, res: *real_t, real_context: *realContext_t) linksection(dr.code_section) void {
    var p: real_t = undefined;

    dr.realDivide(x, b, &p, real_context);
    if (dr.realIsNegative(&p) or dr.realIsZero(&p)) {
        dr.setZero(res);
        return;
    }
    if (dr.realIsSpecial(&p)) {
        dr.setOne(res);
        return;
    }
    dr.realPow(&p, t, &p, real_context);
    dr.realChangeSign(&p);
    dr.WP34S_ExpM1(&p, res, real_context);
    dr.realChangeSign(res);
}

fn wp34sQfWeib(x: *const real_t, b: *const real_t, t: *const real_t, res: *real_t, real_context: *realContext_t) linksection(dr.code_section) void {
    // (-ln(1-p) ^ (1/k)) * J
    var p: real_t = undefined;
    var q: real_t = undefined;

    dr.realMinus(x, &p, real_context);
    dr.WP34S_Ln1P(&p, &p, real_context);
    dr.realChangeSign(&p);
    dr.realDivide(dr.const1(), t, &q, real_context);
    dr.realPow(&p, &q, &p, real_context);
    dr.realMultiply(&p, b, res, real_context);
}
